src/v/cluster/decommissioning_monitor.cc

```diff
@@ -0,0 +1,256 @@
+#include "cluster/decommissioning_monitor.h"
+
+#include "cluster/logger.h"
+#include "cluster/members_table.h"
+#include "cluster/partition_allocator.h"
+#include "cluster/topic_table.h"
+#include "cluster/topics_frontend.h"
+#include "cluster/types.h"
+#include "config/configuration.h"
+#include "model/metadata.h"
+#include "model/timeout_clock.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/util/later.hh>
+
+#include <absl/container/flat_hash_set.h>
+
+#include <algorithm>
+#include <exception>
+#include <iterator>
+
+namespace cluster {
+
+decommissioning_monitor::decommissioning_monitor(
+  ss::sharded<cluster::topics_frontend>& fr,
+  ss::sharded<topic_table>& tt,
+  ss::sharded<partition_allocator>& pal,
+  consensus_ptr raft0,
+  ss::sharded<ss::abort_source>& as)
+  : _topics_frontend(fr)
+  , _topics(tt)
+  , _allocator(pal)
+  , _raft0(raft0)
+  , _as(as)
+  , _retry_timeout(config::shard_local_cfg().decomission_monitor_retry_ms()) {
+    _retry_timer.set_callback([this] {
+        // just signal new changes
+        _new_changes.signal();
+    });
+}
+
+ss::future<> decommissioning_monitor::stop() {
+    _retry_timer.cancel();
+    _new_changes.broken();
+    return _bg.close();
+}
+
+ss::future<> decommissioning_monitor::start() {
+    start_decommission_loop();
+    return ss::now();
+}
+
+void decommissioning_monitor::start_decommission_loop() {
+    return (void)ss::with_gate(_bg, [this] {
+        return ss::do_until(
+          [this] { return _as.local().abort_requested(); },
+          [this] {
+              return _new_changes.wait()
+                .then([this] {
+                    return ss::parallel_for_each(
+                      _reallocations, [this](reallocation_meta& meta) {
+                          return reallocate_replica_set(meta);
+                      });
+                })
+                .then([this] {
+                    // remove finished meta
+                    auto it = std::remove_if(
+                      std::begin(_reallocations),
+                      std::end(_reallocations),
+                      [](reallocation_meta& meta) { return meta.is_finished; });
+
+                    _reallocations.erase(it, std::end(_reallocations));
```

Comment:
this step could be combined with the next or previous continuation, right? it doesn't block